### PR TITLE
Added Blue Score the respective  folders

### LINKED ISSUE #352: Add Bleu score
### Motivation

Add Blue score to future NLP package:
* class implementation can be taken from flash: https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/text/seq2seq/translation/metric.py
* already have a functional version in https://github.com/PyTorchLightning/metrics/blob/master/torchmetrics/functional/nlp.py that just need to be move to a new file in `functional/text/bleu.py` (for consistency)

### Additional context

After merge to torchmetrics + official release the metric should be deprecated from flash + lightning-transformer.

### PR DESCRIPTION
**Pull request by karthikrangasai**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #352 .

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/functional/__init__.py [hunk 1]
```diff
@@ -55,3 +54,4 @@
 from torchmetrics.functional.retrieval.recall import retrieval_recall  # noqa: F401
 from torchmetrics.functional.retrieval.reciprocal_rank import retrieval_reciprocal_rank  # noqa: F401
 from torchmetrics.functional.self_supervised import embedding_similarity  # noqa: F401
+from torchmetrics.functional.text import bleu_score  # noqa: F401
```
Comment (SkafteNicki) [inline torchmetrics/functional/__init__.py:57]: full path :]
```suggestion
from torchmetrics.functional.text.bleu import bleu_score  # noqa: F401
```

File: torchmetrics/functional/__init__.py [hunk 2]
```diff
@@ -33,7 +33,6 @@
 from torchmetrics.functional.classification.specificity import specificity  # noqa: F401
 from torchmetrics.functional.classification.stat_scores import stat_scores  # noqa: F401
 from torchmetrics.functional.image_gradients import image_gradients  # noqa: F401
-from torchmetrics.functional.nlp import bleu_score  # noqa: F401
```
Comment (SkafteNicki) [inline torchmetrics/functional/__init__.py:36]: @Borda should we deprecate this?
Only influences people using the full import path
`from torchmetrics.functional.nlp import bleu_score`
whereas
`from torchmetrics.functional import bleu_score`
will still work

**Review by SkafteNicki — COMMENTED** [review]
Hi @karthikrangasai,
Looks good until now. Could you also:
* Add corresponding tests / move the onces in `tests/functiona/test_nlp.py` to `tests/text/test_bleu.py`
* Add /update references in the docs

**Comment by karthikrangasai** [comment]
@SkafteNicki 
Yeah, I will do the same.

File: torchmetrics/functional/__init__.py (hunk 2 above)
Comment (Borda) [reply torchmetrics/functional/__init__.py:36]: yep, rather make simple deprecation loop

**Comment by karthikrangasai** [comment]
I have refactored the code such that the Class implementation uses the functional implementation of the BLEU score. I have moved the tests for the functional implementation to the respective directory. 

Regarding the tests for the `BLEUScore` class, I have doubts about using the existing `MetricTester` because it accepts `preds` and `targets` as `torch.Tensor` type but the inputs for the BLEU Score metric are an iterable of the strings.

One hack I thought we could use is to provide the iterable of strings as `**mertic_args` and leave `preds` and `targets` as `None` - but this failed all the tests.

Other option I see is to implement a separate `TextMetricTester` for all Text based metrics. I am not too sure about this methods.

How do I proceed with the tests for Class based Text Metrics?

File: torchmetrics/text/bleu.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        super().__init__()
+        self.n_gram = n_gram
+        self.smooth = smooth
+
+        self.add_state("c", tensor(0, dtype=torch.float), dist_reduce_fx="sum")
+        self.add_state("r", tensor(0, dtype=torch.float), dist_reduce_fx="sum")
+        self.add_state("numerator", torch.zeros(self.n_gram), dist_reduce_fx="sum")
+        self.add_state("denominator", torch.zeros(self.n_gram), dist_reduce_fx="sum")
+
+    def update(self, reference_corpus, translate_corpus) -> None:
```
Comment (SkafteNicki) [inline torchmetrics/text/bleu.py:59]: missing typing for `reference_corpus` and `translate_corpus`

File: torchmetrics/text/bleu.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+from torch import tensor
+from torch.functional import Tensor
+
+from torchmetrics import Metric
+from torchmetrics.functional.text.bleu import _bleu_score_compute, _bleu_score_update
+
+
+class BLEUScore(Metric):
+    """
+    Calculate BLEU score of machine translated text with one or more references.
```
Comment (SkafteNicki) [inline torchmetrics/text/bleu.py:29]:
```suggestion
    Calculate `BLEU score <https://en.wikipedia.org/wiki/BLEU>`_ of machine translated text with one or more references.
```

File: torchmetrics/functional/text/bleu.py [hunk 5]
```diff
@@ -115,3 +101,48 @@ def bleu_score(
...
+        raise ValueError(f"Corpus has different size {len(translate_corpus)} != {len(reference_corpus)}")
+    numerator = torch.zeros(n_gram)
+    denominator = torch.zeros(n_gram)
+    c = tensor(0, dtype=torch.float)
+    r = tensor(0, dtype=torch.float)
+
+    c, r = _bleu_score_update(reference_corpus, translate_corpus, numerator, denominator, c, r, n_gram)
+
+    trans_len = c.clone().detach()
+    ref_len = r.clone().detach()
```
Comment (SkafteNicki) [inline torchmetrics/functional/text/bleu.py:146]: why is the `clone().detach()` necessary?

File: torchmetrics/functional/text/bleu.py [hunk 6]
```diff
@@ -97,9 +73,19 @@ def bleu_score(
...
+    return c if isinstance(c, Tensor) else tensor(c), r if isinstance(r, Tensor) else tensor(r)
+
 
+def _bleu_score_compute(
+    trans_len: Tensor,
+    ref_len: Tensor,
+    numerator: Tensor,
+    denominator: Tensor,
+    c: float,
+    r: float,
```
Comment (SkafteNicki) [inline torchmetrics/functional/text/bleu.py:85]:
```suggestion
    c: Tensor,
    r: Tensor,
```

File: torchmetrics/functional/text/bleu.py [hunk 7]
```diff
@@ -45,39 +45,15 @@ def _count_ngram(ngram_input_list: Sequence[str], n_gram: int) -> Counter:
...
-    numerator = torch.zeros(n_gram)
-    denominator = torch.zeros(n_gram)
-    c = 0.0
-    r = 0.0
-
+    translate_corpus: Sequence[Sequence[str]],
+    numerator: Tensor,
+    denominator: Tensor,
+    c: float,
+    r: float,
```
Comment (SkafteNicki) [inline torchmetrics/functional/text/bleu.py:54]:
```suggestion
    c: Tensor,
    r: Tensor,
```

File: torchmetrics/functional/text/bleu.py [hunk 8]
```diff
@@ -115,3 +101,48 @@ def bleu_score(
...
+
+
+def bleu_score(
+    reference_corpus: Sequence[Sequence[Sequence[str]]],
+    translate_corpus: Sequence[Sequence[str]],
+    n_gram: int = 4,
+    smooth: bool = False
+) -> Tensor:
+    """
+    Calculate BLEU score of machine translated text with one or more references
```
Comment (SkafteNicki) [inline torchmetrics/functional/text/bleu.py:113]:
```suggestion
    Calculate `BLEU score <https://en.wikipedia.org/wiki/BLEU>`_ of machine translated text with one or more references
```

File: torchmetrics/text/bleu.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+    Args:
+        n_gram:
+            Gram value ranged from 1 to 4 (Default 4)
+        smooth:
+            Whether or not to apply smoothing – Lin et al. 2004
```
Comment (SkafteNicki) [inline torchmetrics/text/bleu.py:35]: can you insert the reference at the button instead
```suggestion
            Whether or not to apply smoothing – see [2]
```

**Review by SkafteNicki — COMMENTED** [review]
mix of comments

File: torchmetrics/functional/text/bleu.py (hunk 5 above)
Comment (karthikrangasai) [reply torchmetrics/functional/text/bleu.py:146]: The initial implementation had this, I'm not really sure why we are doing it. It just makes a copy that is not considered in the computation graph of the model. We can use the same `c, r` I guess. Please let me know what I should do.

Comment (karthikrangasai) [reply torchmetrics/functional/text/bleu.py:146]: Not doing the `clone().detach()` operation doesn't affect the tests.

Comment (SkafteNicki) [reply torchmetrics/functional/text/bleu.py:146]: then I would remove them :]

File: tests/text/test_blue.py [hunk 10]
```diff
@@ -55,28 +57,67 @@
...
+def test_bleu_score_class(weights, n_gram, smooth_func, smooth):
+    bleu = BLEUScore(n_gram=n_gram, smooth=smooth)
     nltk_output = sentence_bleu(
         [REFERENCE1, REFERENCE2, REFERENCE3],
         HYPOTHESIS1,
         weights=weights,
         smoothing_function=smooth_func,
     )
-    pl_output = bleu_score([HYPOTHESIS1], [[REFERENCE1, REFERENCE2, REFERENCE3]], n_gram=n_gram, smooth=smooth)
+    pl_output = bleu([[REFERENCE1, REFERENCE2, REFERENCE3]], [HYPOTHESIS1])
```
Comment (SkafteNicki) [inline tests/text/test_blue.py:104]: would it be possible for you to expand the test to include multiple batches, like:
```python
for batch in batches:
    blue.update(*batch)
pl_output = blue.compute()
```
so it is also tested that accumulation in the modular metric works

File: torchmetrics/functional/nlp.py [hunk 11]
```diff
@@ -69,49 +48,28 @@ def bleu_score(
         >>> reference_corpus = [['there is a cat on the mat'.split(), 'a cat is on the mat'.split()]]
         >>> bleu_score(translate_corpus, reference_corpus)
         tensor(0.7598)
+
```
Comment (SkafteNicki) [inline torchmetrics/functional/nlp.py:51]: can you include a warning similar to this https://github.com/PyTorchLightning/metrics/blob/a4d2a7a81ba482f28edd30bd76abc3cffbaca548/torchmetrics/functional/regression/mean_relative_error.py#L43-L44 in the docstring

**Review by SkafteNicki — COMMENTED** [review]
minor comments

**Comment by karthikrangasai** [comment]
I have pushed the updated changes.

File: docs/source/references/functional.rst [hunk 12]
```diff
@@ -281,17 +281,6 @@ ssim [func]
 .. autofunction:: torchmetrics.functional.ssim
     :noindex:
 
-
-***
-NLP
```
Comment (Borda) [inline docs/source/references/functional.rst:286]: shall we call it NLP or Text?

File: tests/text/test_blue.py [hunk 13]
```diff
@@ -42,6 +44,11 @@
 LIST_OF_REFERENCES = [[REF1A, REF1B, REF1C], [REF2A]]
 HYPOTHESES = [HYP1, HYP2]
```
Comment (Borda) [inline tests/text/test_blue.py:45]:
```suggestion
LIST_OF_REFERENCES = ((REF1A, REF1B, REF1C), (REF2A))
HYPOTHESES = (HYP1, HYP2)
```
rather immutable, and so rename the `LIST_` too

File: torchmetrics/functional/nlp.py [hunk 14]
```diff
@@ -69,49 +48,31 @@ def bleu_score(
...
-    if min(numerator) == 0.0:
-        return tensor(0.0)
-
-    if smooth:
-        precision_scores = torch.add(numerator, torch.ones(n_gram)) / torch.add(denominator, torch.ones(n_gram))
-        precision_scores[0] = numerator[0] / denominator[0]
-    else:
-        precision_scores = numerator / denominator
+    trans_len = tensor(0, dtype=torch.float)
+    ref_len = tensor(0, dtype=torch.float)
```
Comment (Borda) [inline torchmetrics/functional/nlp.py]: just call the new function here, so it is tiny wrapper

**Comment by karthikrangasai** [comment]
@Borda the tests are failing on Github only because it says list type is not hashable. They run fine on my system.

**Comment by Borda** [comment]
> @Borda the tests are failing on Github only because it says list type is not hashable. They run fine on my system.

and what is your system? I assume list shall not be hashable anywhere as it immutable
https://stackoverflow.com/questions/42203673/in-python-why-is-a-tuple-hashable-but-not-a-list

File: tests/text/test_blue.py (hunk 13 above)
Comment (karthikrangasai) [reply tests/text/test_blue.py:45]: `(REF2A)` doesn't make it a tuple and maintains same datatype, hence I made it `tuple(REF2A)`

**Comment by karthikrangasai** [comment]
Yeah, it was my mistake. I was running `python -m pytest -v tests` but I had to run `python -m pytest -v torchmetrics tests`.
I have fixed the failing tests now. I had forgot to update examples in the docstring and they were failing.

File: docs/source/references/functional.rst (hunk 12 above)
Comment (karthikrangasai) [reply docs/source/references/functional.rst:286]: I am not sure about this. NLP also includes speech processing. So, if we are to add those metrics as well then we can call it NLP.

**Comment by Borda** [comment]
> I have fixed the failing tests now. I had forgot to update examples in the docstring and they were failing.

cool, mind commit your fix? :rabbit:

File: docs/source/references/functional.rst (hunk 12 above)
Comment (Borda) [reply docs/source/references/functional.rst:286]: well, speech processing you mean conversion from audio > text and back, right?
bu then you still shall measure the quality against each independently as your prediction and target are always other audio or text, right? so we can split NLP as text and audio... :rabbit: 
cc: @SkafteNicki @maximsch2

Comment (SkafteNicki) [reply docs/source/references/functional.rst:295]: I prefer text, as the data modality that bleu works on is text, similar to how we have grouped other metrics based on data modality.
Also, this does not matter for the end users, as all modular metrics can just be imported with `from torchmetrics import *` and functional `from torchmetrics.functional import *

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]
